project_1/src/simulator.cpp

```diff
@@ -2,60 +2,86 @@
 }
 
-void NetworkSimulator::StartSimulation() {
+void NetworkSimulator::PacketsSending(bool is_first_iteration) {
     // Each server send cwnd_size packets
     for (ServerSender& server: servers) {
+        // Don't check ACKs on first iteration
+        if (!is_first_iteration && !server.HandleACKs()) {
+            // If server has no ACKs, it doesn't send anything
+            continue;
+        }
         // Create event
         events.emplace_back(server, SEND_DATA, server.GetCWNDSize());
-
-        // Create packets, add them to priority_queue
-        for (std::uint32_t i = 0; i < server.GetCWNDSize(); ++i) {
-            // Calculate estimated delivering time of the packet
-            std::uint32_t estimated_delivering_time = server.GetCurrentTime() + server.GetDistance();
-            // Add packet to the priority_queue
-            packets.emplace(server.GetID(), server_receiver.GetID(), server.GetCurrentTime(), i, estimated_delivering_time);
-            
-            // Update server time
-            server.IncreaseCurrentTime(server.GetDistance());
-        }
-        std::cout << events.back();
+        // Send packets
+        server.SendPackets();
     }
+}
 
+void NetworkSimulator::PacketsReceiving() {
+    std::uint64_t received_data_kb = 0;
+    std::uint32_t received_packets = 0;
     // Create acknowledgements
     while (!packets.empty()) {
+        if (received_data_kb >= bandwidth_bytes) {
+            break;
+        }
+        ++received_packets;
+        received_data_kb += packet_size_bytes;
         // Add ACK event
         events.emplace_back(server_receiver, ACKNOWLEDGEMENT, 1);
-        
-        // Index of server sender in vector
+
+        // Index of server-sender in vector
         std::uint32_t sending_server_index = packets.top().GetSenderID() - 1;
-        
+
         // Send an ACK
-        servers[sending_server_index].AddACK(PacketHeader(0, 
-                                                          packets.top().GetSenderID(), 
-                                                          packets.top().GetEstimatedDeliveryTime(),
-                                                          0, 
-                                                          packets.top().GetEstimatedDeliveryTime() + servers[sending_server_index].GetDistance()));
+        servers[sending_server_index].AddACK(ACK(packets.top().GetSenderID(),
+                                                 packets.top().GetEstimatedDeliveryTime() + servers[sending_server_index].GetDistance()));
 
-        std::cout << events.back();
         packets.pop();
     }
 }
 
+void NetworkSimulator::StartSimulation(std::uint32_t simulation_time_sec) {
+    bool is_first_iteration = true;
+    auto start_time = std::chrono::steady_clock::now();
+    std::uint32_t iteration = 0;
+
+    while (std::chrono::duration_cast<std::chrono::seconds>(std::chrono::steady_clock::now() - start_time).count() < simulation_time_sec) {
```

Comment:
Do we still use the physical time rather then simulated time?

Reply:
We are using physical time to measure time of simulation. It is convenient to run simulation for N seconds.